pkg/backup/item_collector.go

```diff
@@ -296,50 +296,70 @@ func (r *itemCollector) getResourceItems(log logrus.FieldLogger, gv schema.Group
 			labelSelector = metav1.FormatLabelSelector(selector)
 		}
 
-		log.Info("Listing items")
-		unstructuredItems := make([]unstructured.Unstructured, 0)
-
-		if r.pageSize > 0 {
-			// If limit is positive, use a pager to split list over multiple requests
-			// Use Velero's dynamic list function instead of the default
-			listPager := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
-				return resourceClient.List(opts)
-			}))
-			// Use the page size defined in the server config
-			// TODO allow configuration of page buffer size
-			listPager.PageSize = int64(r.pageSize)
-			// Add each item to temporary slice
-			list, paginated, err := listPager.List(context.Background(), metav1.ListOptions{LabelSelector: labelSelector})
-			if err != nil {
-				log.WithError(errors.WithStack(err)).Error("Error listing resources")
-				continue
-			}
-			if !paginated {
-				log.Infof("list for groupResource %s was not paginated", gr)
-			}
-			err = meta.EachListItem(list, func(object runtime.Object) error {
-				u, ok := object.(*unstructured.Unstructured)
-				if !ok {
-					log.WithError(errors.WithStack(fmt.Errorf("expected *unstructured.Unstructured but got %T", u))).Error("unable to understand entry in the list")
-					return fmt.Errorf("expected *unstructured.Unstructured but got %T", u)
-				}
-				unstructuredItems = append(unstructuredItems, *u)
-				return nil
-			})
-			if err != nil {
-				log.WithError(errors.WithStack(err)).Error("unable to understand paginated list")
-				continue
+		var labelSelectors []string
+		if r.backupRequest.Spec.LabelSelectors != nil {
+			for _, s := range r.backupRequest.Spec.LabelSelectors {
+				labelSelectors = append(labelSelectors, metav1.FormatLabelSelector(s))
 			}
 		} else {
-			// If limit is not positive, do not use paging. Instead, request all items at once
-			unstructuredList, err := resourceClient.List(metav1.ListOptions{LabelSelector: labelSelector})
+			labelSelectors = []string{}
+		}
+
+		log.Info("Listing items")
+		unstructuredItems := make([]unstructured.Unstructured, 0)
+		for _, label := range labelSelectors {
+			unstructuredList, err := resourceClient.List(metav1.ListOptions{LabelSelector: label})
 			if err != nil {
 				log.WithError(errors.WithStack(err)).Error("Error listing items")
 				continue
 			}
 			unstructuredItems = append(unstructuredItems, unstructuredList.Items...)
 		}
 
+		// use labelSelector (singular) behavior only when the user has not specified labelSelectors(plural) in backup API
```

Comment:
if you want to apply the labelSelector (singular) after applying labelSelectors (plural), please move the assignment of  labelSelector (singular) closer to the code where it's used.

Reply:
Done, moved the code block